Automating quarkus using gitlab (#1202)
quarkusio/quarkusio.github.io: _posts/2021-11-09-logicdrop-automating-quarkus-with-gitlab.adoc

---
layout: post
title: "Automating Quarkus with GitLab"
date: 2021-11-09
tags: user-story, development-tips, gitlab

> synopsis: "Embracing Quarkus has changed how our developers work and existing processes, especially CI/CD, had to be re-architected to provide a more integrated experience. Every step, from inception, to development through deployment, and finally monitoring, is now orchestrated through GitLab."

Reviewer: ```suggestion
synopsis: "Embracing Quarkus has changed how our developers work and existing processes, especially CI/CD, had to be re-architected to provide a more integrated experience. Every step, from inception, to development through deployment, and finally monitoring, is now orchestrated through GitLab."
```
I think you need to add some context in the synopsis and make it clear it's about your infra at Logicdrop.

Author: Rewrote to:

`
Using Quarkus greatly improved our developers productivity and our CI/CD process needed to be redesigned to catch up. In doing so, we wanted a more universal experience and leveraging Quarkus extensions was a perfect fit. Every step of our Quarkus pipelines, from inception, to development through deployment, and finally monitoring, are now orchestrated through GitLab.
`

Previously, deployments to higher environments were tedious but doable but developers were not fully involved in this process. 

TIP: Since moving to Quarkus, developers are now responsible for developing _and_ deploying (limitations apply) their own services.

We decided to use GitLab https://docs.gitlab.com/ee/ci/pipelines/merge_request_pipelines.html[merge requests pipelines^] as the catalyst to drive our process.

image::gitlab/merge-request.png[Merge Request,align="center"]

On every push, the merge request will:

- Run all the unit tests
- Run any integration tests (Mongo, AWS, Redis)
- Run any E2E tests (other external services)
- Generate code coverage and quality reports
- Publish an interactive Swagger API

*The most useful feature of the merge request is that we can use it to deploy any number of services, and interact with them, into isolated environments.*

image::gitlab/develop-pipeline.png[Review Pipeline,align="center"]

_Review_ services lets us go beyond automated testing. We can spot-test only what is being worked on or even run a full suite of services if needed (this is especially useful during a combined UX and multiple service feature).

> TIP: Lighter than a feather, Quarkus native images are a small fraction of the size compared to their pure Java counterparts. This allows us to deploy a full set of services in less space then would be needed to deploy the equivalent Spring-Boot services.

Reviewer: ```suggestion
TIP: Lighter than a feather, Quarkus native images are a small fraction of the size compared to their pure Java counterparts. This allows us to deploy a full set of services in less space than would be needed to deploy the equivalent Spring Boot services.
```

=== Building Quarkus Services
The downstream Quarkus pipeline is a specialized set of jobs that handle building, testing, and containerizing Quarkus services only.

For each changed service we:

> - Build a native image or FastJar

Reviewer: ```suggestion
- Build a native executable or fast jar
```

- Run any tests (including native ones if needed)
- Generate the Kubernetes manifests
- Build and deploy its container

image::gitlab/develop-downstream.png[Service Downstream Service,align="center"]

TIP: We _only_ push the container and upload the manifests to AWS S3. This lets us accumulate changes into a single package, composed of one or more services, that can be deployed. 

=== Faster Faster Pipelines

> To speed up the pipelines, especially when building multiple native images, we rely on GitLab AutoScaling runners to run the jobs in parallel.  This allows us to build any number of services in a relatively constant time frame.

Reviewer: ```suggestion
To speed up the pipelines, especially when building multiple native executables, we rely on GitLab AutoScaling runners to run the jobs in parallel.  This allows us to build any number of services in a relatively constant time frame.
```
In general, in Quarkus documentation, we prefer the term `native executable` to `native image`. Could you search and replace if it makes sense to you?

Author: Did this across the board.  Thank you for clarifying that.

image::gitlab/view-pods.png[Pod Health,align="center"]

=== Cluster Logs
Kubernetes logs can be viewed right from GitLab, eliminating the need for access to the cluster by developers. 

Logs can be viewed per environment or filtered for specific pods.

image::gitlab/view-logs.png[Cluster Logs,align="center"]

=== Prometheus Metrics
Prometheus metrics are exposed using GitLab's monitoring and metrics. 

image::gitlab/prom.png[Prometheus Metrics,align="center"]

=== Value Stream Analytics
We can use GitLab's https://docs.gitlab.com/ee/user/analytics/value_stream_analytics.html[Value Stream Analytics^] to track the velocity within any project. This gives us rough insights into the time it takes from the moment a ticket is created to when it is deployed.

image::gitlab/valuestream.png[Value Stream Analytics,align="center"]

== Conclusion
Integrating Quarkus deeply into GitLab has added significant value to our process and was well worth the little effort it took. 

Building upon our _level the playing field_ philosophy, developing, debugging, deploying, and monitoring cloud-native services is more optimal than ever.

> Because of the natural fit Quarkus has with cloud technologies, and the functionality provided through extensions, we have been able to create a comprehensive DevOps ecosystem that normally would be challenging to setup and orchestrate.

Reviewer: ```suggestion
Because of the natural fit Quarkus has with cloud technologies, and the functionality provided through extensions, we have been able to create a comprehensive DevOps ecosystem that normally would be challenging to set up and orchestrate.
```

Highlights of the new process are:

- Services can be worked on in isolation
- Parallelizing builds decreased the time to deliver changes
- Common tools are available within GitLab
- Preview services make it easier to collaborate
- Deployments are completely automated

> At the end of the day, we do not have to train staff in different tools, grant access any further than GitLab, or expose any infrastructure directly. 

Reviewer: ```suggestion
At the end of the day, we do not have to train staff on different tools, grant access any further than GitLab, or expose any infrastructure directly. 
```
